### PR TITLE
fix(legend): improve last value handling

### PR DESCRIPTION
**Pull request by markov00**
## Summary

This PR clean up the concept of what the value, shown after the legend item label, represents.

In cartesian charts, the value represents the last available data point laid on the X-axis.
When multiple series are displayed, the last value represents only the data points with the same X as the greatest value on the X axis.

This value, in the future, can also represent another aspect of the series, like the last non-null value, an average/max/min value, but this PR limits is scope to the last value.

We removed every concept of bucket/time interval from this calculation to clarify the legend item value concept.
This allows us to decouple our code from Elasticsearch or other data sources. 
A small, but fixable side-effect is:
- the last value depends entirely on the passed data, so if you pass just 10 values of a series that should cover 100 data points, we consider the last value the last of these 10 values: if the 10th value is not at the extreme edge of your data domain we still consider that 10th value as the last one. If you instead, are looking to depict the extreme edge of your data domain (the 100th element) that in the mentioned example is empty/null then you should pass all the 100 data points with null values where needed.

Related issues
https://github.com/elastic/kibana/issues/153079
https://github.com/elastic/kibana/issues/94280

### BREAKING CHANGE

In cartesian charts, the default legend value now represents the data points that coincide with the latest datum in the X domain. Please consider passing every data point, even the empty ones (like empty buckets/bins/etc) if your x data domain doesn't cover fully a custom x domain passed to the chart configuration.

### Checklist

- [x] The proper **chart type** label has been added (e.g. `:xy`, `:partition`)
- [x] The proper **feature** labels have been added (e.g. `:interactions`, `:axis`)
- [x] All related issues have been linked (i.e. `closes #123`, `fixes #123`)
- [ ] New public API exports have been added to `packages/charts/src/index.ts`
- [x] Unit tests have been added or updated to match the most common scenarios
- [ ] The proper documentation and/or storybook story has been added or updated
- [ ] The code has been checked for cross-browser compatibility (Chrome, Firefox, Safari, Edge)
- [ ] Visual changes have been tested with all available themes including `dark`, `light`, `eui-dark` & `eui-light`

### CODE REVIEW

**Comment by markov00** [comment]
buildkite test this

**Comment by markov00** [comment]
buildkite update screenshots

**Comment by markov00** [comment]
@nickofthyme could you please review this?

File: storybook/stories/test_cases/21_domain_edges.story.tsx [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+          timeAxisLayerCount={2}
+          style={{
+            tickLabel: {
+              alignment: { horizontal: 'left' },
+              padding: 0,
+            },
+            tickLine: {
+              padding: 0,
+              size: 5,
+            },
```
Comment (nickofthyme) [inline storybook/stories/test_cases/21_domain_edges.story.tsx:97]: With the new eui theme adoption, the `tickLine.visible` property is `false`. So with multilayer time axes I add it explicitly.

```suggestion
            tickLine: {
              visible: true,
              padding: 0,
              size: 5,
            },
```

File: storybook/stories/test_cases/21_domain_edges.story.tsx [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
```
Comment (nickofthyme) [inline storybook/stories/test_cases/21_domain_edges.story.tsx:146]: Same thing here

```suggestion
            tickLine: {
              visible: true,
              padding: 0,
              size: 5,
            },
```

File: storybook/stories/test_cases/21_domain_edges.story.tsx [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+          timeAxisLayerCount={2}
+          style={{
+            tickLabel: {
+              alignment: { horizontal: 'center' },
+              padding: 0,
+            },
+            tickLine: {
+              padding: 0,
+              size: 5,
+            },
```
Comment (nickofthyme) [inline storybook/stories/test_cases/21_domain_edges.story.tsx:193]: Same thing here

```suggestion
            tickLine: {
              visible: true,
              padding: 0,
              size: 5,
            },
```

File: packages/charts/src/chart_types/xy_chart/state/utils/get_last_value.ts [hunk 4]
```diff
@@ -6,58 +6,52 @@
...
-    if (series.stackMode === StackMode.Percentage) {
-      const y1InPercentage = y1 === null || y0 === null ? null : y1 - y0;
-      lastValues.set(seriesKey, { y0, y1: y1InPercentage });
-      return;
-    }
-    if (initialY0 !== null || initialY1 !== null) {
-      lastValues.set(seriesKey, { y0: initialY0, y1: initialY1 });
+  switch (type) {
+    case LegendValue.LastNonNullValue: {
+      const last = series.data.findLast((d) => d.x === xDomain.dataDomain[1] && valueAccessor(d) !== null);
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/xy_chart/state/utils/get_last_value.ts:44]: Nice use of `findLast`!

File: packages/charts/src/chart_types/xy_chart/state/utils/get_last_value.ts [hunk 5]
```diff
@@ -6,58 +6,52 @@
...
 import { ScaleType } from '../../../../scales/constants';
 import { XDomain } from '../../domains/types';
 import { isDatumFilled } from '../../rendering/utils';
-import { DataSeries, getSeriesKey, XYChartSeriesIdentifier } from '../../utils/series';
-import { StackMode } from '../../utils/specs';
+import { DataSeries, DataSeriesDatum } from '../../utils/series';
+
+/** @internal */
+export const LegendValue = Object.freeze({
+  None: 'none' as const,
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/xy_chart/state/utils/get_last_value.ts:18]: What's the use of this `None` option, compared to setting `showLegendExtra` to `false`?

**Review by nickofthyme — APPROVED** [review]
Changes LGTM, just one suggestion on the story styles and a question.

File: packages/charts/src/chart_types/xy_chart/state/utils/get_last_value.ts (hunk 5 above)
Comment (markov00) [reply packages/charts/src/chart_types/xy_chart/state/utils/get_last_value.ts:18]: Right now this is only internal, but we will add more legend calculation in the future

File: storybook/stories/test_cases/21_domain_edges.story.tsx (hunk 3 above)
Comment (markov00) [reply storybook/stories/test_cases/21_domain_edges.story.tsx]: fixed in https://github.com/elastic/elastic-charts/pull/2115/commits/6af36e5d71d0adc0fbbb643837888f58df5b35b2

File: storybook/stories/test_cases/21_domain_edges.story.tsx (hunk 2 above)
Comment (markov00) [reply storybook/stories/test_cases/21_domain_edges.story.tsx]: fixed in https://github.com/elastic/elastic-charts/pull/2115/commits/6af36e5d71d0adc0fbbb643837888f58df5b35b2

File: storybook/stories/test_cases/21_domain_edges.story.tsx (hunk 1 above)
Comment (markov00) [reply storybook/stories/test_cases/21_domain_edges.story.tsx]: fixed in https://github.com/elastic/elastic-charts/pull/2115/commits/6af36e5d71d0adc0fbbb643837888f58df5b35b2

**Pull request merged**